Fix numeric conversions (except implicit narrowing/sign-changing integer casts)

stinos opened the pull request:
New version of #5805 with more warnings fixed.

Commits enabling them are kept separate because I'm not sure where we're going with this yet.

On some fairly recent combinations of gcc and glibc this will unfortunately not compile due to a bug in glibc's math.h: the logic for selecting the correct version (depending on whether the argument is float or double) of functions like fpclassify (and sometimes isinf, isnan, ... as well) is broken, resulting in e.g. fpclassify being effectively declared as fpclassify(float) so narrowing conversions from double raise a warning. I cannot pinpoint the exact bug, but search the internet for `"__MATH_TG"` and it shows many similar reports (and fixes to workaround this for clang, which hence is not affected in any recent version I tried).

So currently the Travis build passes, but that's just luck I guess. For gcc 7.5.0 on Ubuntu 18.04 a single fix was sufficient in objint.c:
```
#ifdef __GNUC__
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wfloat-conversion"
#endif
    int cl = fpclassify(val);
#ifdef __GNUC__
#pragma GCC diagnostic pop
#endif
```
But with gcc 9.3.0 on mingw-w64 there's way more problems because isinf etc. are also affected.

What do we do here? Waiting for glibc to fix this is going to take a long time. Adding a bunch of pragmas is also not my favorite. Disabling `-Wfloat-conversion` kinda defeats the purpose of this PR. Custom gcc-specific implementation where we forward to __fpclassify is also a bit messy. Casting the argument to float when needed is also a bit messy. Ideas?

stinos committed:
Revert "all: Fix implicit casts of float/double, and signed comparison."

This reverts commit a2110bd3fca59df8b16a2b5fe4645a4af30b06ed.
There's nothing inherently wrong with it, but upcoming commits will
apply similar fixes in a slightly different way.

stinos committed:
py: Fix sign mismatch in comparison.

Found when compiling with clang and -Wsign-compare.

stinos committed:
all: Fix implicit conversion from double to float.

These are found when using building with -Wfloat-conversion.

stinos committed:
tests/float: Fix cmath_fun_special for MICROPY_FLOAT_IMPL_FLOAT.

When the unix and windows ports use MICROPY_FLOAT_IMPL_FLOAT instead
of MICROPY_FLOAT_IMPL_DOUBLE, the test output has for example
complex(-0.15052, 0.34109) instead of the expected
complex(-0.15051, 0.34109).
Use one decimal place less for the output printing to fix this.

dlech commented:
In "all: Fix implicit floating point promotion" commit message...

> complex(-0.15051, 0.34109) instead of the expected
complex(-0.15051, 0.34109).

Is one of the numbers supposed to be different?

dlech commented:
> What do we do here?

Could we make a macro like `MICROPY_FLOAT_C_FUN` to wrap these that has the appropriate cast on the argument  based on the selected `MICROPY_FLOAT_IMPL`?

```C
#if MICROPY_FLOAT_IMPL == MICROPY_FLOAT_IMPL_DOUBLE
#define MICROPY_MATH_C_FUN(f, a) f((double)a)
...
```

stinos commented:
> Is one of the numbers supposed to be different?

Oops, yes, first one should be `complex(-0.15052, 0.34109)`, will fix.

> Could we make a macro like MICROPY_FLOAT_C_FUN to wrap these that has the appropriate cast 

The problem is that glibc should do something like (pesudocode)
```
#define fpclassify(x) \
if(typeof(x) == double) \
  return __fpclassify(x) \
else \
  return __fpclassifyf(x)
```
but doesn't, so fpclassify gets defined to __fpclassifyf which takes a float as argument. So whatever we put in fpclassify it gets implicitly casted to float, and `f((double)a)` doesn't change that. We could write `fpclassify((float)x)` (in select cases) to get rid of the warning but I'm pretty sure there are edge cases where that gives incorrect results, e.g. very small doubles will get changed into `(float) 0.f` I think which would classify the original number as 0 even when it isn't. Don't know if we should care, but it's not really ok.

Now fpclassify is only used in this particular case here to check whether a number is inf or nan so I could just drop fpclassify and use separate isinf/isnan calls. But that's still not going to compile with latest gcc in mingw.

Other alternative is completely custom implementation of fpclassify, seems we already have mp_fp_as_int_class_t which does part of that. And in formatfloat.c there's also already a custom isinf/isnan/iszero for float, so maybe dpgeorge knows his way around floating point numbers good enough to also do that for double :)

dpgeorge commented on ports/minimal/Makefile:
```diff
@@ -19,12 +19,12 @@ INC += -I$(BUILD)
 else
 LD = gcc
-CFLAGS = -m32 $(INC) -Wall -Werror -std=c99 $(COPT)
+CFLAGS = -m32 $(INC) -Wall -Wdouble-promotion -Werror -std=c99 $(COPT)
```
Does it make sense to enable float warnings on ports like this that don't enable floats at all?

stinos replied:
Not sure, if not enabling flotas means there effectively isn't a single line of code using float or double it doesn't matter of course. But you could also say it doesn't hurt, and it's there already in case someone uses this as a base for another port or just plays around and does enable floats.

Btw I think I just forgot -Wfloat-conversion here.

But in any case: in general do you think this is the way to go, having as much ports as possible use it (some of the warnings found were only because of that) and each makefile specifying it's own flags?

stinos replied:
Also: should the commits enabling the warnings all be squashed, and do we put them before or after the ones with the conversion fixes?

dpgeorge replied:
Definitely it should be up to each port to decide and enable the warnings that it wants to enable (ie these settings shouldn't be centralised).

I guess it makes sense (and doesn't hurt) to have these enabled in those ports that could enable floats.  As you say, as a basis for another port.

> Also: should the commits enabling the warnings all be squashed, and do we put them before or after the ones with the conversion fixes?

Yeah, squashing them seems sensible (so one commit that adds all warnings to all ports/variants).  And putting them after the fixes would be my preference, so that each commit still builds without error (generally try to do this, eg add tests after fixing a bug, but not always possible).

dpgeorge commented on py/nativeglue.c:
```diff
@@ -229,21 +229,13 @@ STATIC bool mp_native_yield_from(mp_obj_t gen, mp_obj_t send_value, mp_obj_t *re
-STATIC mp_obj_t mp_obj_new_float_from_f(float f) {
-    return mp_obj_new_float((mp_float_t)f);
-}
+mp_obj_t mp_obj_new_float_from_f(float f);
```
Is this standard C99, that declaring the function here will add a definition of the static-inline version defined in the header file?

stinos replied:
It will only work if there's another source file including that header, but I do think it's standard. Will try to look it up to make sure though.

stinos replied:
Complicated matter.. Looked around and I did not find a definitive answer. Possibly this should be `extern inline` to force a definition. However nativeglue.c includes runtime.h which includes mpstate.h which includes obj.h so none of this is needed, easy enough :)

dpgeorge replied:
> so none of this is needed,

You mean none of these declarations are needed, and it's legal to take the pointer (in `mp_fun_table`) of a `static inline` function?

stinos replied:
Yes I'm like, 99.9% certain that is required by the standard. The .map file also shows this: nativeglue.o gets the 4 functions locally. 

AFAIK inline means 'compiler, you can inline this if you want', static here means 'compiler, I'm putting this in a header so you can clearly see the definition of what you can inline but it's static because I don't want multiple definitions in case you didn't inline it' and taking the address means 'you cannot inline this thing in this translation unit because then it would not have an address'. However comparing address between different translation units would be wrong but that is not the case here.

I find the standard pretty hard to read though and cannot find the exact place where it says this; Wikipedia explicitly says so though :) But if no-one can confirm we could for safety just introduce locally scoped functions here again.

dpgeorge replied:
Ok, let's just keep it out.  If it breaks on some compiler then it's easy to add them back explicitly.

dpgeorge commented:
This looks good with the from_f/from_d/to_f/to_d functions, thank you!

Regarding the `fpclassify` issue, that is a pain indeed.  Because we already try to support a wide range of compilers (old and new) it might be best to allow full customisation of these FP macros so any issues with the toolchain can be worked around for any given port.

Since `fpclassify` itself is only used in one location we could (as suggested above) either replace it with isinf/isnan, or modify `mp_classify_fp_as_int` to also classify inf/nan.

And if isinf/isnan give problems (I guess only for systems configured to use double?) then we could provide custom versions of them, enabled by a setting like `MICROPY_USE_INTERNAL_ISINF_ISNAN`.

stinos committed:
py: Do not use fpclassify.

For combinations of certain versions of glibc and gcc the definition
of fpclassify always takes float as argument instead of adapting itself
to float/double/long double as required by the C99 standard.
At the time of writing this happens for instance for glibc 2.27
with gcc 7.5.0 when compiled with -Os and glibc 3.0.7 with gcc 9.3.0.
When calling fpclassify with double as argument, as in objint.c, this
results in an implicit narrowing conversion which is not really correct
plus results in a warning when compiled with -Wfloat-conversion so fix
this by spelling out the logic manually.

stinos commented:
I added a commit which removes the use of `fpclassify` and figures out inf/float manually.

Would be better to do this in `mp_classify_fp_as_int`, but then it should be changed to work with `mp_float_union_t` I think, and we probably should look to extend it so it's usable everywhere, should decrease code size since some duplication would be removed, I think. At first sight formatfloat.c and mpz.c have code which could be replaced by `mp_classify_fp_as_int` if it's implementation is more complete (and perhaps takes the `mp_float_union_t` as pointer argument because some code in mpz.c still requires the fraction after classifying what kind of number it is). Would that be feasible?

dpgeorge commented:
> I added a commit which removes the use of `fpclassify` and figures out inf/float manually.

Nice, it's quite simple!  And I just tested the code-size change for this commit alone and it is:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:   -71 -0.014% [incl -8(data)]
unix nanbox:   -59 -0.013% [incl -4(data)]
      stm32:   -12 -0.003% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:    +8 +0.001% GENERIC
      esp32:    +0 +0.000% GENERIC
        nrf:   -28 -0.019% pca10040
```
So I think that's a decent change in it's own right.

> I think, and we probably should look to extend it so it's usable everywhere, should decrease code size since some duplication would be removed, I think

That's a bit out of scope of this PR, which already has a lot of different things in it.  Since this change already reduces code size (see just above) I think it's good progress to just leave it as you have it already, and improve later if possible.

dpgeorge commented:
@stinos are you able to rebase this on master now that the osx Travis job is fixed, and the error string compression is in?

stinos commented:
> rebase this on master

Done, as well as squashing the commits which enable the warnings.

dpgeorge commented:
Code size change of this whole PR is currently:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:   -55 -0.011% [incl -8(data)]
unix nanbox:   -27 -0.006% [incl -4(data)]
      stm32:    -4 -0.001% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:    +8 +0.001% GENERIC
      esp32:   -16 -0.001% GENERIC
        nrf:   -36 -0.025% pca10040
       samd:   -12 -0.012% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
The decrease is mostly due to the change to `fpclassify()`.  But there are other minor changes, probably not worth worrying about.

dpgeorge commented on .travis.yml:
```diff
@@ -184,6 +184,30 @@ jobs:
+        - make ${MAKEOPTS} -C ports/unix CC=clang test
+      after_failure:
+        - tests/run-tests --print-failures
+
```
The number of jobs are growing, which means slower CI builds.  Perhaps these two can be combined into one job?  Would need a "make clean" in between building with gcc and clang.

Also, the clang config does not need the stackless configs set, that was just to test stackless and here should not be set.

stinos replied:
I thought of that already at one point. Combine just these 2, or also the unix stackless one since that also uese clang?

dpgeorge replied:
How about putting the clang 32-bit float into the existing unix-clang job, and the gcc 32-bit float into unix nan-box job?

stinos replied:
Sure.

Btw the number of jobs isn't yet the limiting factor in Travis, it's the OSX build. By the time it effectively starts all other jobs usually finished already..

dpgeorge replied:
If that's the case the osx job should go at the top, to start first.

dpgeorge commented on ports/esp32/Makefile:
```diff
@@ -382,6 +382,8 @@ OBJ_MP += $(addprefix $(BUILD)/, $(EXTMOD_SRC_C:.c=.o))
 OBJ_MP += $(addprefix $(BUILD)/, $(LIB_SRC_C:.c=.o))
 OBJ_MP += $(addprefix $(BUILD)/, $(DRIVERS_SRC_C:.c=.o))
 
+$(OBJ_MP): CFLAGS += -Wdouble-promotion -Wfloat-conversion
```
Does this need to be restricted to OBJ_MP, or can it be applied to global CFLAGS?

stinos replied:
Doubt it: I made separate flags wherever it wasn't doable to apply the flags to the whole source. Don't remember what the problem was here, either libm or the esp-idf SDK. Will check again and add comments.

dpgeorge commented on ports/esp32/espneopixel.c:
```diff
@@ -24,14 +24,14 @@ void IRAM_ATTR esp_neopixel_write(uint8_t pin, uint8_t *pixels, uint32_t numByte
-        period = (fcpu * 2.5) / 1000000; // 1.25us per bit
+        time0 = (uint32_t)((fcpu * 0.5) / 1000000.); // 0.35us
+        time1 = (uint32_t)((fcpu * 1.2) / 1000000.); // 0.8us
+        period = (uint32_t)((fcpu * 2.5) / 1000000.); // 1.25us per bit
```
This code should really be rewritten to use plain ints (it's easy, eg `time0 = fcpu * 350 / 1000`, with `fcpu = ets_get_cpu_frequency()`), but that's out of scope of this commit.  I'd be happy to make this change separately, but it'd need to go before this PR else the build would error.

stinos replied:
Good point, yes please go ahead and change it already.

dpgeorge replied:
See #5888.  I'll need to test then can merge that.

dpgeorge replied:
Ok, that's now merged.

dpgeorge commented on ports/stm32/Makefile:
```diff
@@ -342,6 +342,8 @@ SRC_O += \
 endif
 endif
 
+$(SRC_O): CFLAGS += -Wdouble-promotion -Wfloat-conversion
```
Can these warnings be applied to the global CFLAGS?  Before this commit, with `MICROPY_FLOAT_IMPL=MICROPY_FLOAT_IMPL_FLOAT`, the `Wdouble-promotion` was actually part of the global CFLAGS but now it's not.  So ideally this would be globally applied.

stinos replied:
Same reason as above, but yes the `Wdouble-promotion` should stay where it was.

dpgeorge commented on py/obj.h:
```diff
@@ -775,6 +775,40 @@ bool mp_obj_get_int_maybe(mp_const_obj_t arg, mp_int_t *value);
+static inline mp_obj_t mp_obj_new_float_from_d(double o) {
+    return mp_obj_new_float(o);
+}
+#endif
```
It might be a bit cleaner to put this down below in the `// float` section, before the `#if MICROPY_FLOAT_HIGH_QUALITY_HASH` line.  Otherwise the blank lines here are inconsistent within this block.

dpgeorge commented on py/objfloat.c:
```diff
@@ -52,11 +52,13 @@ typedef struct _mp_obj_float_t {
-const mp_obj_float_t mp_const_float_e_obj = {{&mp_type_float}, M_E};
-const mp_obj_float_t mp_const_float_pi_obj = {{&mp_type_float}, M_PI};
+const mp_obj_float_t mp_const_float_e_obj = {{&mp_type_float}, (mp_float_t)M_E };
+const mp_obj_float_t mp_const_float_pi_obj = {{&mp_type_float}, (mp_float_t)M_PI };
```
The spaces before the `}` here should probably go.

stinos replied:
I'll check if Uncrustify can do that, likewsie for the spaces after `//`

dpgeorge replied:
The option `sp_cmt_cpp_start` will do spaces after the comment, but you'll see that it changes also commented out code.... but I guess that's just a good excuse to remove commented out code!

dpgeorge commented on py/objint.c:
```diff
@@ -135,11 +135,15 @@ STATIC mp_fp_as_int_class_t mp_classify_fp_as_int(mp_float_t val) {
-    } else if (cl == FP_NAN) {
-        mp_raise_ValueError(MP_ERROR_TEXT("can't convert NaN to int"));
+    mp_float_union_t u = {val};
+    //IEEE-754: if biased exponent is all 1 bits...
```
Space just after `//`

dpgeorge commented on py/objint.c:
```diff
@@ -135,11 +135,15 @@ STATIC mp_fp_as_int_class_t mp_classify_fp_as_int(mp_float_t val) {
+    mp_float_union_t u = {val};
+    //IEEE-754: if biased exponent is all 1 bits...
+    if (u.p.exp == ((1 << MP_FLOAT_EXP_BITS) - 1)) {
+        //...then number is Inf (positive or negative) if fraction is 0, else NaN.
```
Space just after `//`

stinos committed:
travis: Decrease build duration by starting OSX build early.

Looking at the recent build history the time it takes just to
complete the OSX build is already 12 minutes so make it start early,
which brings down the total build time from about 20 minutes
to 14 minutes.

stinos committed:
all: Fix implicit floating point promotion.

Initially some of these were found building the unix coverage variant
on MacOS because that build uses clang and has -Wdouble-promotion
enabled, and clang performs more vigorous promotion checks than gcc.
Additionally the codebase has been compiled with clang and msvc (the
latter with warning level 3), and with MICROPY_FLOAT_IMPL_FLOAT to find
the rest of the conversions.
Fixes are implemented either as explicit casts, or by using the correct
type, or by using one of the utility functions to handle floating point
casting; these have been moved from nativeglue.c to the public API.

dpgeorge commented on .travis.yml:
```diff
@@ -168,19 +187,25 @@ jobs:
         - make ${MAKEOPTS} -C ports/unix PYTHON=python2 deplibs
         - make ${MAKEOPTS} -C ports/unix PYTHON=python2 VARIANT=nanbox
         - make ${MAKEOPTS} -C ports/unix PYTHON=python2 VARIANT=nanbox test_full
```
If the test fails the following commands will still execute and the test failure files will be deleted/overwritten.  I think this can be fixed by adding to this line: ` || travis_terminate 1`

dpgeorge commented on .travis.yml:
```diff
@@ -168,19 +187,25 @@ jobs:
         - make ${MAKEOPTS} -C mpy-cross CC=clang
         - make ${MAKEOPTS} -C ports/unix submodules
         - make ${MAKEOPTS} -C ports/unix CC=clang CFLAGS_EXTRA="-DMICROPY_STACKLESS=1 -DMICROPY_STACKLESS_STRICT=1"
         - make ${MAKEOPTS} -C ports/unix CC=clang test
```
same as above `travis_terminate 1`

dpgeorge commented on ports/esp32/Makefile:
```diff
@@ -256,8 +256,9 @@ endif
 # these flags are common to C and C++ compilation
 CFLAGS_COMMON = -Os -ffunction-sections -fdata-sections -fstrict-volatile-bitfields \
 	-mlongcalls -nostdlib \
-	-Wall -Werror -Wno-error=unused-function -Wno-error=unused-but-set-variable \
```
Ok, now I see in the build log that `-Wfloat-conversion` has warnings in the IDF library, so that's why `-Werror` is removed.  In that case it's better to have it the way you had it before with `-Wfloat-conversion` only applied to our code (not IDF).

dpgeorge commented on ports/stm32/Makefile:
```diff
@@ -93,7 +93,7 @@ CFLAGS_MCU_l4 = $(CFLAGS_CORTEX_M) -mtune=cortex-m4 -mcpu=cortex-m4
 CFLAGS_MCU_wb = $(CFLAGS_CORTEX_M) -mtune=cortex-m4 -mcpu=cortex-m4
 
-CFLAGS += $(INC) -Wall -Wpointer-arith -Werror -std=gnu99 -nostdlib $(CFLAGS_MOD) $(CFLAGS_EXTRA)
+CFLAGS += $(INC) -Wall -Wpointer-arith -Wdouble-promotion -Wfloat-conversion -std=gnu99 -nostdlib $(CFLAGS_MOD) $(CFLAGS_EXTRA)
```
As for esp32, `-Werror` will need to stay.  To suppress the resulting warnings/errors might be easy enough to add something like this:
```
$(BUILD)/$(HAL)/Src/%.o: CFLAGS += -Wno-double-promotion
```

stinos replied:
@dpgeorge  Sorry I'm currently in the process to figure out what I can enable and what not, hence the temporary removal of `-Werror`: no time for review yet, should have done this in a separate branch :)

dpgeorge replied:
Ahh, no worries!  Just make a note when it's done.

stinos replied:
Does it help if I click 'Resolve conversation' to indicate an item I've worked on? It's abit confusing because in Gitlab 'Resove conversation' is for the reviewer to indicate something is done, here it's the other way around?

dpgeorge replied:
Yes, you can "resolve conversation" to indicate you dealt with it.  But it's not critical, I will still check all points on final review.  Just make a global comment when this PR is ready to review again.

stinos committed:
all: Enable extra conversion warnings where applicable.

Add -Wdouble-promotion and -Wfloat-conversion for most ports
to ban out implicit floating point conversions and add extra
Travis builds using MICROPY_FLOAT_IMPL_FLOAT to uncover
warnings which weren't found previously.  For the unix port
-Wsign-comparison is added as well but only there since
only clang supports this but gcc doesn't.

stinos committed:
travis: Finish jobs early after test failure.

For jobs which run tests multiple times terminate after the first
run fails otherwise the next test run overwrites the previous
results, making --print-failures useless.

stinos commented:
Up for review again.

Early finish on test failure and making Travis start OSX early are included.

Warnings for libm are now also enabled (qemu-arm port for instance) since fixing them was simply adding `f` to all constants.

Not so for libm-dbl becuase there are a bunch of operations of which I'm not sure how to change (most of it is removing `f` from a bunch of constants however). It would be great though if somebody could check this, just get rid of line 245 in stm32/Makefile.

For the 'Fix implicit floating point to integer conversions.' commit I turned a bunch of floating point arithmetic into integer arithmetic, not 100% sure that is ok.

dpgeorge commented:
> Warnings for libm are now also enabled (qemu-arm port for instance) since fixing them was simply adding `f` to all constants.

This change to libm does look correct, and I diff'd the .o files before and after that change (stm32 build with gcc) and they are equivalent.  So I would be ok with this change to libm.

> Not so for libm-dbl becuase there are a bunch of operations of which I'm not sure how to change (most of it is removing `f` from a bunch of constants however). It would be great though if somebody could check this, just get rid of line 245 in stm32/Makefile.

I had a look at this.  It's possible to fix libm_dbl to build with warnings enabled, and with my changes I even get equivalent .o files before and after (on stm32 build with gcc).  But I don't know if I'd be comfortable making the change because for other compiler it may generate different code.

> For the 'Fix implicit floating point to integer conversions.' commit I turned a bunch of floating point arithmetic into integer arithmetic, not 100% sure that is ok.

Some of these changes are not correct.  I'll comment on them individually.

dpgeorge commented on ports/stm32/machine_timer.c:
```diff
@@ -56,7 +56,7 @@ STATIC mp_obj_t machine_timer_init_helper(machine_timer_obj_t *self, size_t n_ar
         // Frequency specified in Hz
         #if MICROPY_PY_BUILTINS_FLOAT
-        self->delta_ms = 1000 / mp_obj_get_float(args[ARG_freq].u_obj);
+        self->delta_ms = 1000 / (uint32_t)mp_obj_get_float(args[ARG_freq].u_obj);
```
This changes behaviour to do integer instead of float division.  Should be `(uint32_t)(MICROPY_FLOAT_CONST(1000.0) / mp_obj_get_float(...))`.  Hopefully that gives equivalent binaries.

dpgeorge commented on ports/stm32/modusocket.c:
```diff
@@ -319,7 +319,7 @@ STATIC mp_obj_t socket_settimeout(mp_obj_t self_in, mp_obj_t timeout_in) {
     } else {
         #if MICROPY_PY_BUILTINS_FLOAT
-        timeout = 1000 * mp_obj_get_float(timeout_in);
+        timeout = 1000 * (mp_uint_t)mp_obj_get_float(timeout_in);
```
should be a float division

dpgeorge commented on ports/stm32/servo.c:
```diff
@@ -278,7 +278,7 @@ STATIC mp_obj_t pyb_servo_angle(size_t n_args, const mp_obj_t *args) {
     } else {
         #if MICROPY_PY_BUILTINS_FLOAT
-        self->pulse_dest = self->pulse_centre + self->pulse_angle_90 * mp_obj_get_float(args[1]) / 90.0;
+        self->pulse_dest = self->pulse_centre + self->pulse_angle_90 * (uint8_t)mp_obj_get_float(args[1]) / 90;
```
should be a float operation

dpgeorge commented on ports/stm32/servo.c:
```diff
@@ -308,7 +308,7 @@ STATIC mp_obj_t pyb_servo_speed(size_t n_args, const mp_obj_t *args) {
     } else {
         #if MICROPY_PY_BUILTINS_FLOAT
-        self->pulse_dest = self->pulse_centre + self->pulse_speed_100 * mp_obj_get_float(args[1]) / 100.0;
+        self->pulse_dest = self->pulse_centre + self->pulse_speed_100 * (uint8_t)mp_obj_get_float(args[1]) / 100;
```
should be a float operation

dpgeorge commented on ports/stm32/timer.c:
```diff
@@ -281,15 +281,15 @@ STATIC uint32_t compute_prescaler_period_from_freq(pyb_timer_obj_t *self, mp_obj
             freq *= 10;
         }
-        period = (float)source_freq / freq;
+        period = source_freq / (uint32_t)freq;
```
should be float operation

dpgeorge commented on ports/stm32/timer.c:
```diff
@@ -382,7 +382,7 @@ STATIC uint32_t compute_pwm_value_from_percent(uint32_t period, mp_obj_t percent
             cmp = period;
         } else {
-            cmp = percent / 100.0 * ((mp_float_t)period);
+            cmp = period * (uint32_t)percent / 100;
```
should be float operation

stinos committed:
all: Fix implicit floating point to integer conversions.

These are found when using building with -Wfloat-conversion.

stinos commented:
Ok changed back to float operations

dpgeorge commented on extmod/modlwip.c:
```diff
@@ -1316,7 +1316,7 @@ STATIC mp_obj_t lwip_socket_settimeout(mp_obj_t self_in, mp_obj_t timeout_in) {
     } else {
         #if MICROPY_PY_BUILTINS_FLOAT
-        timeout = 1000 * mp_obj_get_float(timeout_in);
+        timeout = 1000 * (mp_uint_t)mp_obj_get_float(timeout_in);
```
this needs to be a float multiplication, because the `timeout_in` may be less than one (to get a sub-second timeout value)

dpgeorge commented on ports/esp32/modsocket.c:
```diff
@@ -457,7 +457,7 @@ STATIC mp_obj_t socket_settimeout(const mp_obj_t arg0, const mp_obj_t arg1) {
     } else {
         #if MICROPY_PY_BUILTINS_FLOAT
-        _socket_settimeout(self, mp_obj_get_float(arg1) * 1000L);
+        _socket_settimeout(self, ((uint64_t)mp_obj_get_float(arg1)) * 1000L);
```
this needs to be a float mult

dpgeorge commented on ports/stm32/servo.c:
```diff
@@ -278,7 +278,7 @@ STATIC mp_obj_t pyb_servo_angle(size_t n_args, const mp_obj_t *args) {
     } else {
         #if MICROPY_PY_BUILTINS_FLOAT
-        self->pulse_dest = self->pulse_centre + self->pulse_angle_90 * mp_obj_get_float(args[1]) / 90.0;
+        self->pulse_dest = self->pulse_centre + self->pulse_angle_90 * (uint16_t)(mp_obj_get_float(args[1]) / MICROPY_FLOAT_CONST(90.0));
```
this needs to be a float mult

dpgeorge commented on ports/stm32/servo.c:
```diff
@@ -308,7 +308,7 @@ STATIC mp_obj_t pyb_servo_speed(size_t n_args, const mp_obj_t *args) {
     } else {
         #if MICROPY_PY_BUILTINS_FLOAT
-        self->pulse_dest = self->pulse_centre + self->pulse_speed_100 * mp_obj_get_float(args[1]) / 100.0;
+        self->pulse_dest = self->pulse_centre + self->pulse_speed_100 * (uint16_t)(mp_obj_get_float(args[1]) / MICROPY_FLOAT_CONST(100.0));
```
this needs to be a float mult

dpgeorge reviewed (commented):
Apart from the 4 lines that need to be changed back to floating point operations (see comments inline), I'm happy with this PR.  (I'm happy to also fix those 4 lines during merge.)

stinos commented:
Fixed

dpgeorge commented:
Ok, merged in 91a6ddc78ada3129bb8968fb0efac1d83e0b99fc through 30840ebc9925bb8ef025dbc2d5982b1bfeb75f1b (I needed to fix one line in `ports/stm32/modusocket.c`, removed the cast to `mp_uint_t` in the call of `mp_obj_get_float()`).

Thanks @stinos for the effort on this!

dpgeorge commented:
@stinos I've just found that old versions of gcc don't support `-Wfloat-conversion`.  Specifically on an old Raspberry Pi, `gcc (Debian 4.6.3-14+rpi1) 4.6.3`.

There could be other compilers/versions that don't like it or other warning options (but hard to find such a definitive list...).  Maybe there needs to be a way to disable these extra warnings, eg by assigning them to a variable in the Makefile like `CWARN_EXTRA` which can be overridden on the command line?

stinos commented:
>  Maybe there needs to be a way to disable these extra warnings, eg by assigning them to a variable in the Makefile like CWARN_EXTRA which can be overridden on the command line?

Yes that would be best. Should all the `-W` flags go in there including `-Werror`?

dpgeorge commented:
> Should all the `-W` flags go in there including `-Werror`?

Well, maybe yes, and then maybe it's fine to just leave it as `CWARN`.  Ie the unix port needs no changes and if the compiler gives bad warnings one can just issue `make CWARN=`.

If that's acceptable then other ports could use a similar mechanism.

stinos commented:
I think that's the best solution for now: it's the easiest, not too much changes required, pretty clear.
